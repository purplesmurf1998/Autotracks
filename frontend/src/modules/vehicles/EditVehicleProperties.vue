<template>
  <div>
    <CCard>
      <CCardBody>
        <CRow class="ml-0 mr-0 d-flex justify-content-between">
          <h4 class="mr-3">Vehicle Properties</h4>
          <div>
            <CButton color="success" @click="submitNewProperties" class="mr-2"
              >Save changes</CButton
            >
            <CButton color="danger" @click="setEditingProperties(false)"
              >Cancel</CButton
            >
          </div>
        </CRow>
        <CAlert v-if="!!errorMessage" color="danger">{{ errorMessage }}</CAlert>
        <hr />
          <CCol v-if="!!editedVehicle">
            <div
              v-for="property in vehicleProperties"
              :key="property._id"
              class="mb-2"
            >
              <CInput
                v-if="property.inputType == 'Text'"
                :name="property.key"
                :label="property.label"
                v-model="editedVehicle[property.key]"
                horizontal
                class="mb-0"
              />
              <CInput
                v-if="property.inputType == 'List'"
                :name="property.key"
                :label="property.label"
                description="Separate list item by semi-colons ';'"
                v-model="editedVehicle[property.key]"
                horizontal
                class="mb-0"
              />
              <CSelect
                v-if="property.inputType == 'Options'"
                :options="property.options"
                :name="property.key"
                :label="property.label"
                :value.sync="editedVehicle[property.key]"
                horizontal
                class="mb-0"
              />
              <CInput
                v-if="property.inputType == 'Date'"
                :label="property.label"
                v-model="editedVehicle[property.key]"
                type="date"
                horizontal
              />
            </div>
          </CCol>
      </CCardBody>
    </CCard>
  </div>
</template>

<script>
const axios = require("axios");

const { getFormattedProperties } = require("../../utils/index");

export default {
  props: [
    "vehicle",
    "leftProperties",
    "rightProperties",
    "vehicleProperties",
    "setEditingProperties",
    "setNewVehicle",
  ],
  data() {
    return {
      editedVehicle: null,
      errorMessage: null
    };
  },
  methods: {
    showError(message) {
      this.errorMessage = message;
      setTimeout(() => {
        this.errorMessage = null;
      }, 5000);
    },
    submitNewProperties() {
      // save the new information
      let properties = this.vehicleProperties;
      let formattedProperties = getFormattedProperties(
        properties,
        this.editedVehicle
      );
      // create the request body
      let tempVehicle = this.vehicle;
      tempVehicle.properties = formattedProperties;
      // update the backend
      this.updateBackend(tempVehicle);
    },
    updateBackend(body) {
      axios({
        method: "PUT",
        url: `${this.$store.state.api}/inventory/vehicle/${this.vehicle._id}`,
        headers: {
          Authorization: `Bearer ${this.$store.state.auth.token}`,
        },
        data: body,
      })
        .then((response) => {
          if (response.data.success) {
            // do something
            this.setNewVehicle(response.data.payload);
            this.setEditingProperties(false);
          }
        })
        .catch((err) => {
          console.log(err);
          this.showError("Error occured while editing vehicle property.");
        });
    },
  },
  mounted() {
    // get the values of the properties and add to the editedVehicle field
    const editedVehicle = {};
    if (this.vehicleProperties) {
      this.vehicleProperties.forEach((property) => {
        if (property.inputType == "List") {
          let value = "";
          let items = this.vehicle.properties[property.key];
          items.forEach((item, index) => {
            if (index < items.length - 1) {
              value += item + ";";
            } else {
              value += item;
            }
          });
          editedVehicle[property.key] = value;
        } else {
          editedVehicle[property.key] = this.vehicle.properties[property.key];
        }
      });
      this.editedVehicle = editedVehicle;
    }
  },
  components: {},
};
</script>

<style>
</style>