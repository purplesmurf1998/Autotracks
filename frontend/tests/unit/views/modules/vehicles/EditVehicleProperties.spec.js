import {createLocalVue, shallowMount} from '@vue/test-utils'
import CoreuiVue from '@coreui/vue-pro'
import Vuex from 'vuex'

import EditVehicleProperties from './../../../../../src/modules/vehicles/EditVehicleProperties.vue'

const localVue = createLocalVue()
localVue.use(CoreuiVue)
localVue.use(Vuex)

const store = new Vuex.Store({
    state: {
      sidebarShow: 'responsive',
      sidebarMinimize: false,
      asideShow: false,
      darkMode: false
    }
  })

  describe('EditVehicleProperties.vue', () => {
    const wrapper = shallowMount(EditVehicleProperties, { 
      store, 
      localVue
    })
    it('has a name', () => {
      expect(EditVehicleProperties.name).toBe('EditVehicleProperties')
    })
    it('should render correct content', () => {
        expect(wrapper.find('h4').text()).toMatch('Vehicle Properties')
    })
    test('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })