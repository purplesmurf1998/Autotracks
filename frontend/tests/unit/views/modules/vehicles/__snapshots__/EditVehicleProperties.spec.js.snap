// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditVehicleProperties.vue renders correctly 1`] = `
<div>
  <ccard-stub>
    <ccardbody-stub>
      <crow-stub
        class="ml-0 mr-0 d-flex justify-content-between"
        gutters="true"
        tag="div"
      >
        <h4
          class="mr-3"
        >
          Vehicle Properties
        </h4>
         
        <div>
          <cbutton-stub
            activeclass="router-link-active"
            class="mr-2"
            color="success"
            event="click"
            exactactiveclass="router-link-exact-active"
            routertag="a"
            target="_self"
            type="button"
          >
            Save changes
          </cbutton-stub>
           
          <cbutton-stub
            activeclass="router-link-active"
            color="danger"
            event="click"
            exactactiveclass="router-link-exact-active"
            routertag="a"
            target="_self"
            type="button"
          >
            Cancel
          </cbutton-stub>
        </div>
      </crow-stub>
       
      <!---->
       
      <hr />
       
      <!---->
    </ccardbody-stub>
  </ccard-stub>
</div>
`;
